rasa/core/evaluation/markers.py

```diff
@@ -0,0 +1,120 @@
+        return {}  # currently represented as a simple dict. Might change later.
+
+    @classmethod
+    def load_config_from_path(cls, path: Union[Text, Path]) -> Dict:
+        """Loads the config from a file or directory."""
+        path = os.path.abspath(path)
+        if os.path.isfile(path):
+            config = cls.from_file(path)
+        elif os.path.isdir(path):
+            config = cls.from_directory(path)
+        else:
+            raise InvalidMarkersConfig(
+                "Failed to load markers configuration from '{}'. "
+                "File not found!".format(os.path.abspath(path))
+            )
+        return config
+
+    @classmethod
+    def from_file(cls, path: Text) -> Dict:
+        """Loads the config from a YAML file."""
+        return cls.from_yaml(rasa.shared.utils.io.read_file(path), path)
+
+    @classmethod
+    def from_yaml(cls, yaml: Text, original_filename: Text = "") -> Dict:
+        """Loads the config from YAML text after validating it."""
+        try:
+            # TODO implement and call the yaml schema validation
+            #  rasa.shared.utils.validation.validate_yaml_schema(yaml,
+            #  MAKERS_SCHEMA_FILE)
+
+            config = rasa.shared.utils.io.read_yaml(yaml)
+            return config
+        except YamlException as e:
+            e.filename = original_filename
+            raise e
+
+    @classmethod
+    def from_directory(cls, path: Text) -> Dict:
+        """Loads and appends multiple configs from a directory tree."""
+        config = cls.empty_config()
+        for root, _, files in os.walk(path, followlinks=True):
+            for file in files:
+                full_path = os.path.join(root, file)
+                if cls.is_markers_config_file(full_path):
+                    other = cls.from_file(full_path)
+                    config = cls._merge(config, other)
+        return config
+
+    @classmethod
+    def _merge(cls, config_a: Dict, config_b: Dict) -> Dict:
+        """Merges multiple marker configs."""
+        copy_config_a = config_a.copy()
+        if "markers" in copy_config_a.keys():
+            copy_config_a["markers"].extend(config_b["markers"])
+            return copy_config_a
+        else:
+            return config_b
+
+    @staticmethod
+    def is_markers_config_file(filename: Text) -> bool:
+        """Checks whether the given file path is a config file.
+
+        Args:
+            filename: Path of the file which should be checked.
+
+        Returns:
+            `True` if it's a config file, otherwise `False`.
+
+        Raises:
+            YamlException: raised when the file has a YAML extension but
+                cannot be read / parsed.
+        """
+        from rasa.shared.data import is_likely_yaml_file
+
+        if not is_likely_yaml_file(filename):
+            return False
+
+        try:
+            # check if the file can be read as a yaml file.
+            rasa.shared.utils.io.read_yaml_file(filename)
```

Comment:
Yes, I agree with you. I based this on Domain.py to try an be consistent with the way we process yaml input, but I too didn't like that the yaml file was being read twice. I think the original author wanted to separated checking that the file is a valid yaml from the process of reading it. I can fold them into a single function if consistency with other parts of the codebase is not a huge concern. 
I yet don't know if the function will be used elsewhere. Did you want me to prepend an underscore to the name?

Reply:
So looking into the Domain one, they have tests to ensure invalid domains are rejected, and otherwise the only place it's really used is for MultiProjectImporter. I'm not sure if that's something that would need extended by this feature? 

I think the tests are a good idea but the multiple reads bothers me a little bit - I suppose it's not critical in this case since it's dealing with a fairly small file, and we can leave it consistent with the rest of the codebase. Might be a possible optimisation thing in future if it comes to that (since it probably wouldn't be hard to attempt a read then return file content if successful else warn)

Reply:
Yeah no worries! I moved stuff around, so now I check if the file has the correct extension, and if so, I process it, raising an exception inside read_yaml if it can't be parsed. This way we get the same exception and error message whether we read multiple configs or just one.